# UFCFVQ-15-M Programming for Data Science
# Programming Task 1

## Student Id: 202111341

### Requirement FR1.1 - Find the arithmetic mean



In [5]:
# add code here
# Read the Glocuse data from task1.dat
task1 = [i.strip() for i in open("task1.dat").readlines()]

# coverting the string into integers
data_task1 = (list(map(int, task1)))

# finding the mean
mean = sum(data_task1) / len(data_task1)

print("There are a total of", len(data_task1), "observations in the Glocuse data.")
print("The arithmetic mean of Glocuse is", mean,".")

There are a total of 768 observations in the Glocuse data.
The arithmetic mean of Glocuse is 120.89453125 .


### Requirement FR1.2 - Find the standard deviation

In [6]:
# add code here
# computing for the variance and standard deviation
variance = sum([((x - mean) ** 2) for x in data_task1]) / len(data_task1)
std = variance ** 0.5
print("The standard deviation of Glocuse is", std)

The standard deviation of Glocuse is 31.95179590820272


### Requirement FR1.3 - Find the min/max values 

In [7]:
# add code here
# Maximum and Minimum values of data
max(data_task1)
min(data_task1)

Min = min(data_task1)
Max = max(data_task1)
print("The Minimum value of Glocuse is", Min,"and the Maximum value is",Max)

The Minimum value of Glocuse is 0 and the Maximum value is 199


### Requirement FR1.4 - Find the 25th percentile 

In [11]:
# add code here
# Compute for the 25th percentile
import math
def my_percentile(a, percentile):
    p = len(data_task1) * percentile / 100
    if p.is_integer():
        return sorted(data_task1)[int(p)]
    else:
        return sorted(data_task1)[int(math.ceil(p)) - 1]
index25 = [25]
p25 = [my_percentile(data_task1, i) for i in index25]
print("The 25th percentile is", p25)

The 25th percentile is [99]


### Requirement FR1.5 - Find the 50th percentile

In [12]:
# add code here
# Compute for the 50th percentile
index50 = [50]
p50 = [my_percentile(data_task1, i) for i in index50]
print("The 50th percentile is", p50)

The 50th percentile is [117]


### Requirement FR1.6 - Find the 75th percentile

In [140]:
# add code here
# Compute for the 75th percentile
index75 = [75]
p75 = [my_percentile(data_task1, i) for i in index75]
print("The 75th percentile is", p75)

The 75th percentile is [141]


### Requirement FR1.7 - Print the set of summary statistics in tabular form

In [14]:
table_data = [
    ['Statistics', 'Value'],
    ['Count', len(data_task1)],
    ['Arithmetic Mean', mean], 
    ['Standard Deviation', std],
    ['Minimum', Min],
    ['25th Percentile', 99],
    ['50th Percentile', 117],
    ['75th Percentile', 141],
    ['Maximum', Max]
]
for row in table_data:
    print("{: >20} {: >20}".format(*row))

          Statistics                Value
               Count                  768
     Arithmetic Mean         120.89453125
  Standard Deviation    31.95179590820272
             Minimum                    0
     25th Percentile                   99
     50th Percentile                  117
     75th Percentile                  141
             Maximum                  199


### Requirement FR1.8 - Read data from a file into memory

In [160]:
# add code here
# Open function to open the file "task1.dat"
# This is already done in FR1.1 above.
task_file1 = [i.strip() for i in open("task1.dat").readlines()]
print(task_file1)

['148', '85', '183', '89', '137', '116', '78', '115', '197', '125', '110', '168', '139', '189', '166', '100', '118', '107', '103', '115', '126', '99', '196', '119', '143', '125', '147', '97', '145', '117', '109', '158', '88', '92', '122', '103', '138', '102', '90', '111', '180', '133', '106', '171', '159', '180', '146', '71', '103', '105', '103', '101', '88', '176', '150', '73', '187', '100', '146', '105', '84', '133', '44', '141', '114', '99', '109', '109', '95', '146', '100', '139', '126', '129', '79', '0', '62', '95', '131', '112', '113', '74', '83', '101', '137', '110', '106', '100', '136', '107', '80', '123', '81', '134', '142', '144', '92', '71', '93', '122', '163', '151', '125', '81', '85', '126', '96', '144', '83', '95', '171', '155', '89', '76', '160', '146', '124', '78', '97', '99', '162', '111', '107', '132', '113', '88', '120', '118', '117', '105', '173', '122', '170', '84', '96', '125', '100', '93', '129', '105', '128', '106', '108', '108', '154', '102', '57', '106', '147'

### Requirement FR1.9 - Design and build a data structure to store CSV data in memory

In [174]:
# add code here
# store CSV data in memory
import csv 
def read_file_csv():
    task1_data = open("task1.csv", "r") # opening the file
    content = csv.reader(myfile) # storing the data in a variable
    return content

### Requirement FR1.10 - Create and demonstrate a function to read CSV data from a file into memory

In [168]:
# add code here
# opening the CSV file
with open('task1.csv', mode ='r') as file:

# reading the CSV file
    task1_csv = csv.DictReader(file)

# displaying the contents of the CSV file
    for lines in task1_csv:
        print(lines)


{'Pregnancies': '6', 'Glucose': '148', 'BloodPressure': '72', 'SkinThickness': '35', 'Insulin': '0', 'BMI': '33.6', 'DiabetesPedigreeFunction': '0.627', 'Age': '50', 'Outcome': '1'}
{'Pregnancies': '1', 'Glucose': '85', 'BloodPressure': '66', 'SkinThickness': '29', 'Insulin': '0', 'BMI': '26.6', 'DiabetesPedigreeFunction': '0.351', 'Age': '31', 'Outcome': '0'}
{'Pregnancies': '8', 'Glucose': '183', 'BloodPressure': '64', 'SkinThickness': '0', 'Insulin': '0', 'BMI': '23.3', 'DiabetesPedigreeFunction': '0.672', 'Age': '32', 'Outcome': '1'}
{'Pregnancies': '1', 'Glucose': '89', 'BloodPressure': '66', 'SkinThickness': '23', 'Insulin': '94', 'BMI': '28.1', 'DiabetesPedigreeFunction': '0.167', 'Age': '21', 'Outcome': '0'}
{'Pregnancies': '0', 'Glucose': '137', 'BloodPressure': '40', 'SkinThickness': '35', 'Insulin': '168', 'BMI': '43.1', 'DiabetesPedigreeFunction': '2.288', 'Age': '33', 'Outcome': '1'}
{'Pregnancies': '5', 'Glucose': '116', 'BloodPressure': '74', 'SkinThickness': '0', 'Insul

### Requirement FR1.11 - Create and demonstrate a generic statistical summary function 

In [ ]:
# add code here
# Opening the CSV file
with open('task1.csv') as csvf: 
    rdata_task1 = csv.reader(csvf, delimiter=',') 
    lines = 0
        for row in csvr: # for loop to loop through the data
        if lines == 0:
            lines +=1
        else:
            data = [float(i) for i in row] # calculate the stats of the data for each lines in the file
            mean(rdata_task1)
            std(rdata_task1)
            Min(rdata_task1)
            p25(rdata_task1)
            p50(rdata_task1)
            p75(rdata_task1)
            Max(rdata_task1)

### Requirement FR1.12 - Create and demonstrate a function to print a custom table

In [ ]:
# add code here
data = task1_csv
table_data = [
    ['Statistics', 'Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'BMI', 'DiabetesPedigreeFunction', 'Age'],
    ['Count', len(data.Pregnancies), len(data.Glucose), len(data.BloodPressure), len(data.SkinThickness), len(data.BMI), len(data.DiabetesPedigreeFunction), len(data.Age)],
    ['Arithmetic Mean', mean(data.Pregnancies),mean(data.Glucose), mean(data.BloodPressure), mean(data.SkinThickness), mean(data.BMI), mean(data.DiabetesPedigreeFunction), mean(data.Age)], 
    ['Standard Deviation', std(data.Pregnancies), std(data.Glucose), std(data.BloodPressure), std(data.SkinThickness), std(data.BMI), std(data.DiabetesPedigreeFunction), std(data.Age)],
    ['Minimum', Min(data.Pregnancies), Min(data.Glucose), Min(data.BloodPressure), Min(data.SkinThickness), Min(data.BMI), Min(data.DiabetesPedigreeFunction), Min(data.Age)],
    ['25th Percentile', p25(data.Pregnancies), p25(data.Glucose), p25(data.BloodPressure), p25(data.SkinThickness), p25(data.BMI), p25(data.DiabetesPedigreeFunction), p25(data.Age)],
    ['50th Percentile', p50(data.Pregnancies), p50(data.Glucose), p50(data.BloodPressure), p50(data.SkinThickness), p50(data.BMI), p50(data.DiabetesPedigreeFunction), p50(data.Age)],
    ['75th Percentile', p75(data.Pregnancies), p75(data.Glucose), p75(data.BloodPressure), p75(data.SkinThickness), p75(data.BMI), p75(data.DiabetesPedigreeFunction), p75(data.Age)],
    ['Maximum', Max(data.Pregnancies), Max(data.Glucose), Max(data.BloodPressure), Max(data.SkinThickness), Max(data.BMI), Max(data.DiabetesPedigreeFunction), Max(data.Age)]
]
for row in table_data:
    print("{: >20} {: >20}".format(*row))

# Process Development Report for Task 1


    For main purpose of this task is to calculate standard statistical measures for the given data st using built-in functions and data structures only and without the use of any library functions. This is quite challenging for students like me whose without prior exprience and next-to-zero knowledge in programming.
    Since it is mentioned in the task description that the two provided data is to be used in this task, I started with reading the data 'task1.dat' file provided. As we are NOT allowed to use any library for this task, finding the code to read the data was daunting. First, data was treated as string and read and stored into a DataFrame. After then, the data was converted into integers to be processed in the next step. This data by the way, as compared to the second file 'task1.csv' is about the 'Glocuse' measure. I was able to extract 768 observations. 
    For the standard statistical measures, we were asked to get the arithmetic mean, the standard deviation, the minimum and maximum values and the 25th, 50th, 75th percentiles. The average (arithmetic mean) was then computed by dividing the sum of the obervations by the number of observations. From the computed mean, I then computed for the variance by taking the squared difference of the mean to all 768 observations and dividing the sum by the total observations. The square-root of the variance gave me the required standard deviations. Finding the minimum and maximum values were pretty straightforward using the min() and max() functions. The problem then followed in computing for the percentiles. In here, I imported the math library to do the calculations. After getting all the required percentile values, it was then presented as a table showing the summary statistics and their corresponding values.
    For the next requirements, FR1.8 is already done in FR1.1, as I used the same data to compute for some of the basic statistical functions. A data structure was then designed to read and store the values of CSV file 'task1.csv'. This was done using the open() and read() functions, as well as the csv.DictReader() file reader function.
    Using the standard statistical functions in FR1.1 to FR1.6, the arithmetic mean, the standard deviation, the minimum and maximum values and the 25th, 50th, 75th percentiles of the CSV data was then computed. Finally, a function to create a custom table that summarizes all the statistics was then created. Multiple errors was experienced in doing this and the final code is of no different.
    In summary, reading, exploring and analyzing data using only the built-in commands of Python was so difficult and challenging. There is a need for data scientists to use libraries to simplify the work and to come up with reliable processing and correct analysis of big data.
    